task/buildah-rhtap/0.1/buildah-rhtap.yaml

```diff
@@ -0,0 +1,237 @@
+      value: $(params.IMAGE_EXPIRES_AFTER)
+  steps:
+  - image: $(params.BUILDER_IMAGE)
+    name: build
+    computeResources:
+      limits:
+        memory: 4Gi
+      requests:
+        memory: 512Mi
+        cpu: 250m
+    env:
+    - name: COMMIT_SHA
+      value: $(params.COMMIT_SHA)
+    script: |
+      SOURCE_CODE_DIR=source
+      if [ -e "$SOURCE_CODE_DIR/$CONTEXT/$DOCKERFILE" ]; then
+        dockerfile_path="$SOURCE_CODE_DIR/$CONTEXT/$DOCKERFILE"
+      elif [ -e "$SOURCE_CODE_DIR/$DOCKERFILE" ]; then
+        dockerfile_path="$SOURCE_CODE_DIR/$DOCKERFILE"
+      elif echo "$DOCKERFILE" | grep -q "^https\?://"; then
+        echo "Fetch Dockerfile from $DOCKERFILE"
+        dockerfile_path=$(mktemp --suffix=-Dockerfile)
+        http_code=$(curl -s -L -w "%{http_code}" --output "$dockerfile_path" "$DOCKERFILE")
+        if [ $http_code != 200 ]; then
+          echo "No Dockerfile is fetched. Server responds $http_code"
+          exit 1
+        fi
+        http_code=$(curl -s -L -w "%{http_code}" --output "$dockerfile_path.dockerignore.tmp" "$DOCKERFILE.dockerignore")
+        if [ $http_code = 200 ]; then
+          echo "Fetched .dockerignore from $DOCKERFILE.dockerignore"
+          mv "$dockerfile_path.dockerignore.tmp" $SOURCE_CODE_DIR/$CONTEXT/.dockerignore
+        fi
+      else
+        echo "Cannot find Dockerfile $DOCKERFILE"
+        exit 1
+      fi
+
+      # Fixing group permission on /var/lib/containers
+      chown root:root /var/lib/containers
+
+      sed -i 's/^\s*short-name-mode\s*=\s*.*/short-name-mode = "disabled"/' /etc/containers/registries.conf
+
+      LABELS=(
+        "--label" "build-date=$(date -u +'%Y-%m-%dT%H:%M:%S')"
+        "--label" "architecture=$(uname -m)"
+        "--label" "vcs-type=git"
+      )
+      [ -n "$COMMIT_SHA" ] && LABELS+=("--label" "vcs-ref=$COMMIT_SHA")
+      [ -n "$IMAGE_EXPIRES_AFTER" ] && LABELS+=("--label" "quay.expires-after=$IMAGE_EXPIRES_AFTER")
+
+      buildah build \
+        ${LABELS[@]} \
+        --tls-verify=$TLSVERIFY --no-cache \
+        --ulimit nofile=4096:4096 \
+        -f "$dockerfile_path" -t $IMAGE $SOURCE_CODE_DIR/$CONTEXT
+
+      container=$(buildah from --pull-never $IMAGE)
+      buildah mount $container | tee /workspace/container_path
+      echo $container > /workspace/container_name
+    securityContext:
+      capabilities:
+        add:
+          - SETFCAP
+    volumeMounts:
+    - mountPath: /var/lib/containers
+      name: varlibcontainers
+    workingDir: $(workspaces.source.path)
+
+  - name: sbom-syft-generate
+    image: quay.io/redhat-appstudio/syft:v0.98.0@sha256:4d3856e6a2622700b9a9d5d74d9aaf5d8a55671653f80bf6c636677658680ede
+    script: |
+      syft dir:$(workspaces.source.path)/source --output cyclonedx-json=$(workspaces.source.path)/sbom-source.json
+      find $(cat /workspace/container_path) -xtype l -delete
+      syft dir:$(cat /workspace/container_path) --output cyclonedx-json=$(workspaces.source.path)/sbom-image.json
```

Comment:
Ok, this took me some time to figure out. I think we pontentially have two approaches to solve this:
- Try to use `oci-dir` instead of `dir`. I couldn't figure out a way to get/mount the image as an oci-dir though
- Manually set the catalogers. I added a patch that does that (in a very inelegant way)

Reply:
For the latter approach, since we still use0.98, we can't use tags to set catalogers (introduced in [0.101](https://github.com/anchore/syft/tree/v0.101.0)), such as:
```
syft <some container image> --select-catalogers "image"
```

So in the commit I added, I needed to manually specify the catalogers as describe in the [0.98 docs](https://github.com/anchore/syft/tree/v0.98.0#image-scanning). Dumb me is still struggling with generating a config yaml file from within the script block from within the task yaml file :laughing:

Any hints are appreciated, but for now, I added a very ugly single-liner that solves the problem (apparently, syft also does not like when we break the args in multiple lines using \, I'm still not sure why).

Reply:
> Try to use oci-dir instead of dir. I couldn't figure out a way to get/mount the image as an oci-dir though

Would this work?

```shell
buildah push "$image" oci-archive:image.tar
syft oci-archive:image.tar
```

Reply:
`oci-archive:` does not do it, but `oci-dir:` works. For reference, these were the catalogers executed this time:
- alpm-db-cataloger
- apk-db-cataloger
- binary-cataloger
- conan-info-cataloger
- dpkg-db-cataloger
- dotnet-portable-executable-cataloger
- go-module-binary-cataloger
- java-archive-cataloger
- graalvm-native-image-cataloger
- javascript-package-cataloger
- nix-store-cataloger
- php-composer-installed-cataloger
- portage-cataloger
- python-installed-package-cataloger
- r-package-cataloger
- rpm-db-cataloger
- ruby-installed-gemspec-cataloger
- sbom-cataloger